### PR TITLE
Date picker improvements

### LINKED ISSUE #1171: Date picker a11y issues
Screen reader test, investigate and fix the various issue with pikaday:

image

### PR DESCRIPTION
**Pull request by jamesjacobs**
This changes adds the option for the new accessible[ Pulsar Date Picker ](https://www.npmjs.com/package/pulsar-date-picker) to be used with the `form.date()` helper.

When used, the new date picker brings a range of accessibility improvements to screen reader experience, keyboard accessibility and more.

For backwards compatibility, Pikaday will remain the default date picker for the short term.

image

Browser tested in:
- Chrome
- Safari
- Firefox
- Edge
- IE11

Accessibility tested in:
- JAWS 2018
- JAWS 18
- NVDA
- Site Improve
- WAVE
- AXE

Fixes #1171

### CODE REVIEW

File: js/area/signin/signin.js [hunk 1]
```diff
@@ -217,7 +217,9 @@ SignInComponent.prototype.reset = function () {
 		.removeClass('signin--error')
 		.find('.signin__hint').remove();
 
-	this.$html.find('input:not([name="username"])').trigger('blur');
+	if (this.$html.find('.signin-container input:not([name="username"])').length) {
```
Comment (Stanton) [inline js/area/signin/signin.js:220]: Why are there changes to the signin component here?

File: views/pulsar/v2/helpers/form.html.twig [hunk 2]
```diff
@@ -901,33 +902,67 @@ show-label  | bool   | Control visibility of the `<label>` element without affec
...
 {% macro date(options) %}
 {% spaceless %}
     {% import _self as form %}
+    {% import '@pulsar/pulsar/v2/helpers/html.html.twig' as html %}
+
+    {# check for use of pulsar-date-picker #}
+    {% if options['data-pulsar-datepicker'] is defined and options['data-pulsar-datepicker'] is not empty and options['data-pulsar-datepicker'] == true %}
+
+        {# Generate guid to link input to trigger button, required for pulsar-date-picker #}
+        {% set guidDatePickerTriggerGuid = 'guid-' ~ random() %}
```
Comment (Stanton) [inline views/pulsar/v2/helpers/form.html.twig:931]: Minor, but the duplication of guid in the var could be simplified? `datePickerTriggerGuid` ?

File: views/pulsar/v2/helpers/form.html.twig [hunk 3]
```diff
@@ -901,33 +902,67 @@ show-label  | bool   | Control visibility of the `<label>` element without affec
...
+            {% set isDisabled = true %}
+        {% endif %}
+
+        {% set dateOptions = {
+            'placeholder': options.placeholder|default('dd/mm/yyyy'),
+            'data-pulsar-datepicker-format': options.format|default,
+            'data-pulsar-datepicker': options['data-pulsar-datepicker']|default('true'),
+            'data-pulsar-datepicker-trigger': guidDatePickerTriggerGuid,
+            'append': html.button({
+                'label': html.icon('calendar', { 'label': 'Show calendar' }),
```
Comment (Stanton) [inline views/pulsar/v2/helpers/form.html.twig:945]: You're happy this label will work with CXM translations?

File: js/area/signin/signin.js (hunk 1 above)
Comment (jamesjacobs) [reply js/area/signin/signin.js:220]: That method (`SignInComponent.reset`) was bound to a `ESC` keyup event on `document` and was triggering every time `ESC` was pressed on any input, which interfered with focus management on the date picker (and potentially other things that listen for `ESC`). I've tried to keep the fix as simple as possible by only triggering blur on inputs within the `signin-container`.

File: views/pulsar/v2/helpers/form.html.twig (hunk 2 above)
Comment (jamesjacobs) [reply views/pulsar/v2/helpers/form.html.twig]: Yep typo, good spot, fixed

**Pull request merged**